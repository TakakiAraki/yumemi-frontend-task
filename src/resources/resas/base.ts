import { RESAS_API_KEY } from "~/utils/env";
import { RequestError } from "~/errors/api/RequestError";
import ResponseError from "~/errors/api/ResponseError";
import { clock } from "~/utils/clock";
import { promiseControl, waitFor } from "~/utils/promise";

const BASE_URL = "https://opendata.resas-portal.go.jp";

/**
 * リクエスト数/日 10000
 * 1秒あたりのリクエスト平均数 5
 * @param url
 * @param init
 * @returns
 */
export const fetchResas = async (
  url: RequestInfo | URL,
  init?: RequestInit,
): Promise<Response | RequestError | ResponseError> => {
  try {
    const result = await fetch(BASE_URL + url, {
      ...init,
      headers: {
        ...init?.headers,
        "X-API-KEY": RESAS_API_KEY,
      },
    });
    if (!result.ok) return new ResponseError(result);
    return result;
  } catch (_error) {
    return new RequestError(url, init);
  }
};
