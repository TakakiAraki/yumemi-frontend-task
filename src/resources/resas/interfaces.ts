import { ResourceResult } from "interfaces/resources";

export interface Prefectures {
  message?: any;
  result: PrefecturesResult[];
}

interface PrefecturesResult {
  prefCode: number;
  prefName: string;
}

export type PrefecturesAPIRequestProps<T = unknown> = T;
export type PrefecturesAPIResult = ResourceResult<Prefectures>;

export interface Demographics {
  message?: any;
  result: Result;
}

interface DemographicsProps {
  prefCode: string;
  cityCode: string; // 「すべての市区町村」を選択する場合は「-」を送ります。
  addArea?: {
    prefCode: string;
    cityCode: string;
  }[]; // addArea=都道府県コード_市区町村コード addArea=1_01100,13_13101
}

interface Result {
  boundaryYear: number;
  data: Datum2[];
}

interface Datum2 {
  label: string;
  data: Datum[];
}

interface Datum {
  year: number;
  value: number;
  rate?: number;
}

export type DemographicsAPIRequestProps = DemographicsProps;
export type DemographicsAPIResult = ResourceResult<Demographics>;
