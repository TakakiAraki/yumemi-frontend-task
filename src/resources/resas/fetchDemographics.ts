import { RequestError } from "~/errors/api/RequestError";
import ResponseError from "~/errors/api/ResponseError";
import { is } from "~/utils/Is";
import { fetchResas } from "./base";
import {
  DemographicsAPIRequestProps,
  DemographicsRepository,
} from "~/usecases/demographics/interface";

// prefCode: string;
// cityCode: string;
export const parseAddArea = (addArea: Required<DemographicsAPIRequestProps>["addArea"]): string =>
  addArea.map((area) => `${area.cityCode}_${area.prefCode}`).join(",");

/**
 * https://opendata.resas-portal.go.jp/docs/api/v1/population/composition/perYear.html
 */

export const fetchDemographics: DemographicsRepository = async (props) => {
  const { addArea, ...params } = props;

  const param = new URLSearchParams(params);
  if (is.notNull(addArea)) param.set("addArea", parseAddArea(addArea));
  const response = await fetchResas("/api/v1/population/composition/perYear?" + param.toString());

  if (response instanceof RequestError || response instanceof ResponseError) {
    return {
      state: "ERROR",
      result: {
        code: response.name,
        message: response.message,
      },
    };
  }
  const result = await response.json();
  return {
    state: "SUCCESS",
    result: result,
  };
};
