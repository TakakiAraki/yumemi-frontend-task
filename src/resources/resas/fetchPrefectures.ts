import { RequestError } from "errors/api/RequestError";
import ResponseError from "errors/api/ResponseError";
import { fetchResas } from "./base";
import { PrefecturesAPIResult } from "./interfaces";

/**
 * https://opendata.resas-portal.go.jp/docs/api/v1/detail/index.html
 */
export const fetchPrefectures = async (): PrefecturesAPIResult => {
  const response = await fetchResas("/api/v1/prefectures");

  if (response instanceof RequestError) {
    return {
      state: "ERROR",
      result: {
        code: response.name,
        message: response.message,
      },
    };
  }

  if (!response.ok) {
    const result = new ResponseError(response);
    return {
      state: "ERROR",
      result: {
        code: result.name,
        message: result.message,
      },
    };
  }

  const result = await response.json();
  return {
    state: "SUCCESS",
    result: result,
  };
};
