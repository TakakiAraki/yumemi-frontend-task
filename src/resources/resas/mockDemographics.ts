import { DemographicsRepository } from "~/usecases/demographics/interface";
import mock from "./demographics.json";

export const mockDemographics: DemographicsRepository = async (_) => {
  // TODO: テストケースを記述する際はMockを増やしprefCodeなどを利用する
  return await {
    state: "SUCCESS",
    result: mock,
  };
};
