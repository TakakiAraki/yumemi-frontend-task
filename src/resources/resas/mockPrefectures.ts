import { PrefecturesRepository } from "~/usecases/prefectures/interfaces";
import mock from "./prefectures.json";

export const mockPrefectures: PrefecturesRepository = async () => {
  return {
    state: "SUCCESS",
    result: mock,
  };
};
