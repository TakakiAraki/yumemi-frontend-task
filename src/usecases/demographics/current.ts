import { fetchDemographics } from "~/resources/resas/fetchDemographics";
import { reduces } from "~/utils/array";
import { is } from "~/utils/Is";
import { DemographicsRepository } from "./interface";

interface ICurrent {
  prefCode: string;
  cityCode?: string;
}
interface IRepository {
  api: DemographicsRepository;
}

export default async ({ prefCode, cityCode = "-" }: ICurrent, repository?: IRepository) => {
  const api = repository?.api || fetchDemographics;

  const { state, result } = await api({
    prefCode,
    cityCode,
  });

  if (state === "ERROR") {
    return {
      state,
      result,
    };
  }

  const boundaryYear = result.result.boundaryYear;
  const data = result.result.data.find((prefectureData) => prefectureData.label !== "総人口");
  const value = data?.data.reduce(
    reduces.filterMap((population) => {
      if (population.year > boundaryYear) return false;
      return {
        date: population.year,
        values: {
          [prefCode]: population.value,
        },
      };
    }),
    [],
  );

  if (is.null(value)) {
    // TODO: ErrorObjectを作成

    return {
      state: "ERROR",
      result: {
        code: "",
        message: "",
      },
    };
  }

  return {
    state,
    result: value,
  };
};
