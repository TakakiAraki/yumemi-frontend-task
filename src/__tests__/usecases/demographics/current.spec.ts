import { mockDemographics } from "~/resources/resas/mockDemographics";
import current from "~/usecases/demographics/currentDemographics";

describe("usecase test - demographics", () => {
  test("current test", (done) => {
    current({ prefCode: "1" }, { api: mockDemographics }).then((res) => {
      expect(res).toMatchSnapshot();
      done();
    });
  });
});
