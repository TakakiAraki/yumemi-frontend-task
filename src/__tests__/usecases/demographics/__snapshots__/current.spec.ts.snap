// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usecase test - demographics current test 1`] = `
{
  "result": [
    {
      "date": 1960,
      "values": {
        "1": 1681479,
      },
    },
    {
      "date": 1965,
      "values": {
        "1": 1462123,
      },
    },
    {
      "date": 1970,
      "values": {
        "1": 1309487,
      },
    },
    {
      "date": 1975,
      "values": {
        "1": 1312611,
      },
    },
    {
      "date": 1980,
      "values": {
        "1": 1298324,
      },
    },
    {
      "date": 1985,
      "values": {
        "1": 1217959,
      },
    },
    {
      "date": 1990,
      "values": {
        "1": 1034251,
      },
    },
    {
      "date": 1995,
      "values": {
        "1": 898673,
      },
    },
    {
      "date": 2000,
      "values": {
        "1": 792352,
      },
    },
    {
      "date": 2005,
      "values": {
        "1": 719057,
      },
    },
    {
      "date": 2010,
      "values": {
        "1": 657312,
      },
    },
    {
      "date": 2015,
      "values": {
        "1": 608296,
      },
    },
  ],
  "state": "SUCCESS",
}
`;
